### PR TITLE
Enhancement: Have `fetchAndCompile` return name of fetcher used

### PR DESCRIPTION
**Pull request by haltman-at**
Addresses #4446.

The `SingleRecognizer` now records and returns the name of the fetcher used, in addition to the other information.  Note that to do this I put `fetchedVia` into the `FetchAndCompileResult` type; so I removed the third argument from `addCompiledInfo` as it was now redundant.  I think we can reasonably say that's not really a breaking change, as it doesn't break either `fetchAndCompile` or `fetchAndCompileForDebugger`; it only breaks `fetchAndCompileForRecognizer`, which... c'mon, that's pretty internal (it's not even used elsewhere in Truffle although I did export it).

The functionality of `fetchAndCompileForDebugger` remains unchanged.

If we really are worried about that sort of thing, I can redo it in a truly non-breaking way, but this was the most straightforward way.

Note this PR has no tests, because fetch-and-compile has no tests; but once #4424 is merged I can go back and add tests of this.  (Or if @leeftk merges this in then he can add the tests on his PR, but I don't know that he wants that extra work. XD )

### CODE REVIEW

**Comment by gnidan** [comment]
@haltman-at it might be a bit of extra work for you to maintain the branch this way, but if you rebase this against #4424 then you and @leeftk could collaborate to get tests here in this PR!

**Review by gnidan — APPROVED** [review]
Approving this without tests because this might end up being a blocker on the tx-insight side otherwise

**Review by cds-amal — COMMENTED** [review]
I'd prefer if we had tests for this but @gnidan approved! So it's your call @haltman-at . Perhaps the tests could be add on work after #4424 is merged?

**Comment by haltman-at** [comment]
Yes, that's my intention.

**Pull request merged**
